tooling/minitest/src/tests/builder_api.rs

```diff
@@ -0,0 +1,115 @@
+use crate::*;
+
+#[test]
+fn global_var() {
+    let mut p = ProgramBuilder::new();
+    let var = p.global_int::<u32>();
+    let mut f = p.function();
+    f.assign(var, const_int(42u32));
+    f.if_(eq(load(var), const_int(42u32)), |_| {}, |f| f.unreachable());
+    f.exit();
+    p.finish_function(f);
+    let p = p.finish_program();
+    assert_stop(p);
+}
+
+#[test]
+fn local_var() {
+    let mut p = ProgramBuilder::new();
+    let mut f = p.function();
+    let var = f.local::<u32>();
+    f.storage_live(var);
+    f.assign(var, const_int(42u32));
+    f.if_(eq(load(var), const_int(42u32)), |_| {}, |f| f.unreachable());
+    f.exit();
+    p.finish_function(f);
+    let p = p.finish_program();
+    assert_stop(p);
+}
+
+#[test]
+fn arg_and_ret_var() {
+    let mut p = ProgramBuilder::new();
+    let mut f = p.function();
+    let var = f.arg::<u32>();
+    let ret = f.ret::<u32>();
+    f.if_(eq(load(var), const_int(42u32)), |_| {}, |f| f.unreachable());
+    f.assign(ret, add(load(var), const_int(2u32)));
+    f.exit();
+    let add_two_if_42 = p.finish_function(f);
+
+    let mut start = p.function();
+
+    let l = start.local::<u32>();
+    start.storage_live(l);
+    start.call(add_two_if_42, &[by_value(const_int(42u32))], l, Ret::Yes);
+    start.if_(eq(load(l), const_int(44u32)), |f| f.exit(), |f| f.unreachable());
+    let start = p.finish_function(start);
+
+    p.set_start(start);
+    let p = p.finish_program();
+    assert_stop(p);
+}
+#[test]
+#[should_panic(expected = "PlaceExpr is not a local")]
+fn storage_live_with_non_local() {
+    let mut p = ProgramBuilder::new();
+    let g = p.global_int::<i32>();
+    let mut f = p.function();
+    f.storage_live(g);
+}
+
+#[test]
+#[should_panic(expected = "PlaceExpr is not a local")]
+fn storage_dead_with_non_local() {
+    let mut p = ProgramBuilder::new();
+    let g = p.global_int::<i32>();
+    let mut f = p.function();
+    f.storage_dead(g);
+}
+
+#[test]
+fn switch_int() {
+    let mut p = ProgramBuilder::new();
+    let mut f = p.function();
+
+    let var = f.local::<u32>();
+    f.storage_live(var);
+    f.assign(var, const_int(42_u32));
+    f.switch_int(
+        load(var),
+        &[
+            (41, &|f| f.assign(var, add(load(var), const_int(0u32)))),
+            (42, &|f| f.assign(var, add(load(var), const_int(1u32)))),
+            (43, &|f| f.assign(var, add(load(var), const_int(0u32)))),
+        ],
+        |f| f.assign(var, add(load(var), const_int(0u32))),
+    );
+    f.if_(eq(load(var), const_int(43_u32)), |_| {}, |f| f.unreachable());
+    f.storage_dead(var);
+    f.exit();
```

Comment:
What happens if you don't exit here?

Reply:
Added a test for this case &rarr; should now throw a nice error msg.